tests/checker/httpserver.py

```diff
@@ -175,19 +177,45 @@ def get_url(self, filename):
         return u"http://localhost:%d/tests/checker/data/%s" % (self.port, filename)
 
 
+class HttpsServerTest(HttpServerTest):
+    """
+    Start/stop an HTTPS server that can be used for testing.
+    """
+
+    def setUp(self):
+        """Start a new HTTPS server in a new thread."""
+        self.port = start_server(self.handler, https=True)
+        assert self.port is not None
+
+    def tearDown(self):
+        """Send QUIT request to http server."""
+        stop_server(self.port, https=True)
+
+    def get_url(self, filename):
+        """Get HTTP URL for filename."""
+        return u"https://localhost:%d/tests/checker/data/%s" % (self.port, filename)
+
 
-def start_server (handler):
+def start_server (handler, https=False):
     """Start an HTTP server thread and return its port number."""
     server_address = ('localhost', 0)
     handler.protocol_version = "HTTP/1.0"
     httpd = StoppableHttpServer(server_address, handler)
+    if https:
+        httpd.socket = ssl.wrap_socket(httpd.socket,
+            keyfile=get_file("https_key.pem"),
+            certfile=get_file("https_cert.pem"), server_side=True)
     port = httpd.server_port
     t = threading.Thread(None, httpd.serve_forever)
     t.start()
     # wait for server to start up
     while True:
         try:
-            conn = HTTPConnection("localhost:%d" % port)
+            if https:
+                conn = HTTPSConnection("localhost:%d" % port,
+                        context=ssl._create_unverified_context())
```

Comment:
Is that a documented API method?  I worry because it starts with an underscore.

Reply:
_create_unverified_context() is not mentioned in https://docs.python.org/3/library/ssl.html.  Let's not use it.

We can inline the implementation from the stdlib and eliminate all the constants from default arguments, which will give us

```python
context = ssl.SSLContext()
context.check_hostname = False
context.verify_mode = ssl.CERT_NONE
```

and the docs say that check_hostname is False by default except in one specific circumstance that doesn't apply to us, so this assignment can be omitted.  The default value of verify_mode is not documented.  I'm pretty sure it is CERT_NONE in all current versions of Python, so we might simplify this to

```python
                conn = HTTPSConnection(..., context=ssl.SSLContext())
```

Reply:
> _create_unverified_context() is not mentioned in https://docs.python.org/3/library/ssl.html. Let's not use it.

It came from the http.client.HTTPSConnection documentation:
https://docs.python.org/3/library/http.client.html?highlight=_create_unverified_context#http.client.HTTPSConnection

but I don't mind replacing it.

Reply:
That makes me feel much better, but also makes me want to SCREAM at Python developers who did this.